addons/web/controllers/home.py

```diff
@@ -110,7 +111,11 @@ def web_login(self, redirect=None, **kw):
 
         if request.httprequest.method == 'POST':
             try:
-                uid = request.session.authenticate(request.db, request.params['login'], request.params['password'])
+                credential = {'type': 'password'}
+                for cred_type in CREDENTIAL_PARAMS:
+                    if cred_type in request.params:
+                        credential[cred_type] = request.params[cred_type]
+                uid = request.session.authenticate(request.db, credential)
```

Comment:
This is completely nitpicking but I would write this with a one-liner. This is one of the cases I believe a dict comprehension is clearer code than a loop.
```py
credential = {key: value for key, value in request.params.items() if key in CREDENTIAL_PARAMS}
credential['type'] = 'password'
```